### PR TITLE
Add args to after_initialize hooks for JSON and YAML serialization

### PR DESCRIPTION
**Pull request by dukeraphaelng**
- Resolves https://github.com/crystal-lang/crystal/issues/10265

### CODE REVIEW

File: src/json/serialization.cr [hunk 1]
```diff
@@ -134,13 +134,13 @@ module JSON
       # Define a `new` directly in the included type,
       # so it overloads well with other possible initializes
 
-      def self.new(pull : ::JSON::PullParser)
-        new_from_json_pull_parser(pull)
+      def self.new(pull : ::JSON::PullParser, **args)
```
Comment (Sija) [inline src/json/serialization.cr:137]: _args_ is typically used with single splats, I'd suggest sticking with _opts_ instead.

```suggestion
      def self.new(pull : ::JSON::PullParser, **opts)
```

ditto elsewhere

**Comment by Blacksmoke16** [comment]
I'm not so sure this is the best approach.  These overloads aren't really meant to be called manually.  It feels kinda hacky IMO.  Could you not do something like:

```cr
class JSONAttrPersonWithYAMLInitializeHookWithParams
  include JSON::Serializable
  include YAML::Serializable

  property name : String
  property age : Int32
  property trusted : Bool?

  def self.from_trusted_json(string_or_io : String | IO) : self
    obj = self.from_json string_or_io
    obj.trusted = true
    obj
  end

  def self.from_trusted_yaml(string_or_io : String | IO) : self
    obj = self.from_yaml string_or_io
    obj.trusted = true
    obj
  end

  def initialize(@name : String, @age : Int32)
  end
end
```

**Comment by dukeraphaelng** [comment]
@Blacksmoke16 

I'm sorry that was really not the best example. I'm upgrading an ORM at the moment to migrate from JSON.mapping, and YAML.mapping to JSON and YAML Serializable.

Essentially, I want to call this following method:

``` Crystal
    def after_initialize(trusted : Bool)
      if !trusted
        {% for name, opts in FIELDS %}
          {% if !opts[:mass_assign] %}
            @{{name}} = nil
          {% end %}
        {% end %}
      end

      apply_defaults
      clear_changes_information
    end
```

In `def clear_changes_information`, I want to modify the instance variables directly and not through using a setter. Because I only want users to use setter to change acceptable fields, and not metadata fields of the class.

However, modifying an instance variables outside of a `def initialize` is illegal. Hence I thought this would be the appropriate solution.

**Comment by j8r** [comment]
You can set a `protected` setter.
What about having two structs, including/inheriting from an other one, say `TrustedPersonEtc` and `PersonEtc`?
All the logic will be there, and the interface will be `abstract def trusted : Bool`.

**Comment by straight-shoota** [comment]
> However, modifying an instance variables outside of a def initialize is illegal.

Of course you can just assign instance variables from *any* instance method without requiring a setter. In fact, a setter is just an ordinary method that assigns to an ivar.

I don't think this change is necessary, it just adds a lot of complexity without a reasonable benefit. There's a simple workaround as explained in https://github.com/crystal-lang/crystal/issues/10265#issuecomment-762240658
`after_initialize` should just be a simple argless hook.

**Comment by dukeraphaelng** [comment]
Thanks a lot for all of your inputs! Especially @straight-shoota , this has resolved my issue! I have been working on this forever, really glad I was able to resolve this with helpful responses from everyone.

**Comment by straight-shoota** [comment]
Feel free to reach out for help earlier via chat or forum ;)

**Pull request closed without merging**